openrtb-json/src/main/scala/com/powerspace/openrtb/json/bidrequest/OpenRtbNativeRequestSerde.scala

```diff
@@ -1,48 +1,88 @@
 
-object OpenRtbNativeRequestSerde extends EncoderProvider[Native] with NativeDependencies {
   import io.circe._
   import io.circe.syntax._
   import EncodingUtils._
-  import com.powerspace.openrtb.json.common.OpenRtbProtobufEnumEncoders._
-
-  import io.circe.generic.extras.semiauto._
-
-  private implicit val titleEncoder: Encoder[NativeRequest.Asset.Title] = openrtbEncoder[Asset.Title]
-  private implicit val imgEncoder: Encoder[NativeRequest.Asset.Image] = openrtbEncoder[Asset.Image]
+  import OpenRtbProtobufEnumEncoders._
+  import OpenRtbProtobufEnumDecoders._
 
-  private implicit val videoEncoder = OpenRtbVideoSerde.encoder
-
-  private implicit val assetDataEncoder: Encoder[NativeRequest.Asset.Data] = openrtbEncoder[Asset.Data].cleanRtb
+  private implicit val titleEncoder: Encoder[NativeRequest.Asset.Title] = openRtbEncoder[Asset.Title]
+  private implicit val imgEncoder: Encoder[NativeRequest.Asset.Image] = openRtbEncoder[Asset.Image]
+  private implicit val videoEncoder: Encoder[Imp.Video] = OpenRtbVideoSerde.encoder
+  private implicit val assetDataEncoder: Encoder[NativeRequest.Asset.Data] = openRtbEncoder[Asset.Data].cleanRtb
   private implicit val assetOneOfEncoder: Encoder[NativeRequest.Asset.AssetOneof] = protobufOneofEncoder[NativeRequest.Asset.AssetOneof] {
     case NativeRequest.Asset.AssetOneof.Img(img) => img.asJson
     case NativeRequest.Asset.AssetOneof.Data(data) => data.asJson
     case NativeRequest.Asset.AssetOneof.Video(video) => video.asJson
     case NativeRequest.Asset.AssetOneof.Title(title) => title.asJson
   }
 
-  private implicit val assetEncoder: Encoder[NativeRequest.Asset] = openrtbEncoder[Asset]
-  private implicit val eventTrackersEncoder: Encoder[NativeRequest.EventTrackers] = openrtbEncoder[EventTrackers]
+  private implicit val assetEncoder: Encoder[NativeRequest.Asset] = openRtbEncoder[Asset]
+  private implicit val eventTrackersEncoder: Encoder[NativeRequest.EventTrackers] = openRtbEncoder[EventTrackers]
+  implicit val nativeRequestEncoder: Encoder[NativeRequest] = openRtbEncoder[NativeRequest]
 
   private implicit val requestOneOfEncoder: Encoder[Imp.Native.RequestOneof] = protobufOneofEncoder[Imp.Native.RequestOneof] {
     case Imp.Native.RequestOneof.Request(string) => string.asJson
     case Imp.Native.RequestOneof.RequestNative(request) => request.asJson
   }
 
-  implicit val nativeRequestEncoder: Encoder[NativeRequest] = openrtbEncoder[NativeRequest]
+  def encoder: Encoder[Imp.Native] = openRtbEncoder[Imp.Native]
+
+  private implicit val titleDecoder: Decoder[NativeRequest.Asset.Title] = openRtbDecoder[Asset.Title]
+  private implicit val imgDecoder: Decoder[NativeRequest.Asset.Image] = openRtbDecoder[Asset.Image]
+  private implicit val videoDecoder: Decoder[Imp.Video] = OpenRtbVideoSerde.decoder
+  private implicit val assetDataDecoder: Decoder[NativeRequest.Asset.Data] = openRtbDecoder[Asset.Data]
+
+  private implicit val assetDecoder: Decoder[NativeRequest.Asset] = cursor =>
```

Comment:
Can it be refactored the `.prepare` way (just like the BidRequest)  ?

Reply:
I'll try it out, if it's too tricky I'll leave it as it is.